### PR TITLE
Dependency fix on typescript CLI

### PR DESCRIPTION
**Pull request by reebayroo**
This PR contains the following: 

- Fixing the support for Http and File dependencies on the `dependencies` property of morphir.json file
- Fixing the `localDependencies` compatibility
- Integration tests for local and dependencies 
- Unit tests for the dependencies module

# Terms

> THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE TERMS OF THE CCLA DATED 2017-11-07 WITH FINOS/LINUX FOUNDATION (FORMERLY THE SYMPHONY SOFTWARE FOUNDATION CCLA).
>
> THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.

### CODE REVIEW

File: cli2/dependencies.test.ts
Comment (reebayroo) [inline cli2/dependencies.test.ts:1]: Unit test for the dependencies file

File: cli2/dependencies.ts [hunk 1]
```diff
@@ -9,7 +11,7 @@ import { ResultAsync } from "neverthrow";
 const parseDataUrl = require("data-urls");
 const fsReadFile = util.promisify(fs.readFile);
 
-const DataUrl = z.string().trim().transform((val, ctx) => {
+export const DataUrl = z.string().trim().transform((val, ctx) => {
```
Comment (reebayroo) [inline cli2/dependencies.ts:14]: Exposing this objects so they can be unit tested

File: cli2/dependencies.ts [hunk 2]
```diff
@@ -21,7 +23,7 @@ const DataUrl = z.string().trim().transform((val, ctx) => {
   return parsed;
 });
 
-const FileUrl = z.string().trim().url().transform((val, ctx) => {
+export const FileUrl = z.string().trim().url().transform((val, ctx) => {
```
Comment (reebayroo) [inline cli2/dependencies.ts:26]: Exposing this objects so they can be unit tested

File: cli2/dependencies.ts [hunk 3]
```diff
@@ -32,19 +34,34 @@ const FileUrl = z.string().trim().url().transform((val, ctx) => {
...
+  }
+  console.error(`${fullPath} does not exist`);
+  ctx.addIssue({
+    code: z.ZodIssueCode.custom,
+    message: `File not found ${val}`
+  })
+  return z.NEVER;
+})
+
+export const Url = z.string().url().transform((url) => new URL(url));
```
Comment (reebayroo) [inline cli2/dependencies.ts:52]: Creating a Local file to support compatibility

File: cli2/dependencies.ts [hunk 4]
```diff
@@ -32,19 +34,34 @@ const FileUrl = z.string().trim().url().transform((val, ctx) => {
...
 
 const PathOrUrl = z.union([FileUrl, z.string().trim().min(1)]);
 
-const GithubData = z.object({
+export const GithubData = z.object({
```
Comment (reebayroo) [inline cli2/dependencies.ts:55]: Exposing this objects so they can be unit tested

File: cli2/dependencies.ts [hunk 5]
```diff
@@ -32,19 +34,34 @@ const FileUrl = z.string().trim().url().transform((val, ctx) => {
...
   repo: z.string(),
   baseUrl: z.string().optional()
 });
 
 const GithubConfig = z.union([GithubData, z.string()]);
 
-const DependencySettings = z.union([DataUrl, FileUrl, z.string().trim()])
+// const DependencySettings = z.union([DataUrl, FileUrl, z.string().trim()])
+const DependencySettings =  z.string().trim();
 const Dependencies = z.array(DependencySettings).default([]);
```
Comment (reebayroo) [inline cli2/dependencies.ts:65]: This is the fix for the Dependencies

File: cli2/dependencies.ts [hunk 6]
```diff
@@ -112,7 +129,7 @@ export async function loadAllDependencies(config: DependencyConfig) {
   const finalResults = await Promise.all(results);
   return finalResults.flatMap((result) => {
     if (result.isOk()) {
-      console.error("Successfully loaded dependency", result.value.dependency)
+      console.info("Successfully loaded dependency", result.value.dependency)
```
Comment (reebayroo) [inline cli2/dependencies.ts:132]: removing error messages for debugging purposes

File: cli2/cli.ts [hunk 7]
```diff
@@ -46,7 +46,8 @@ async function make(
   const dependencyConfig = DependencyConfig.parse({
     dependencies: morphirJson.dependencies,
     localDependencies: morphirJson.localDependencies,
-    includes: includes
+    includes: includes, 
+    projectDir: projectDir
   })
```
Comment (reebayroo) [inline cli2/cli.ts:51]: ProjectDir preserves the project directory option passed through the command line

File: cli2/dependencies.ts [hunk 8]
```diff
@@ -1,15 +1,17 @@
...
-import { getUri } from "get-uri";
+import { ResultAsync } from "neverthrow";
+import * as path from "path";
+import * as util from "util";
 import { decode, labelToName } from "whatwg-encoding";
+import { z } from "zod";
+import { fetchUriToJson } from "./get-uri-wrapper";
 import { Readable } from "stream";
-import { ResultAsync } from "neverthrow";
+
```
Comment (reebayroo) [inline cli2/dependencies.ts:9]: Organizing imports

File: cli2/dependencies.ts [hunk 9]
```diff
@@ -32,25 +34,50 @@ const FileUrl = z.string().trim().url().transform((val, ctx) => {
...
 
-const Url = z.string().url().transform((url) => new URL(url));
+type LocalFileRef = {
+  projectDir: string, 
+  original: string,
+  fullPath: string,
+  url?: URL,
+
+}
+
```
Comment (reebayroo) [inline cli2/dependencies.ts:44]: Used by LocalFile to keep track of the different forms of a path

File: cli2/dependencies.ts [hunk 10]
```diff
@@ -1,15 +1,17 @@
-import * as util from "util";
 import * as fs from "fs";
-import { z } from "zod";
-import { getUri } from "get-uri";
+import { ResultAsync } from "neverthrow";
+import * as path from "path";
+import * as util from "util";
 import { decode, labelToName } from "whatwg-encoding";
+import { z } from "zod";
+import { fetchUriToJson } from "./get-uri-wrapper";
```
Comment (reebayroo) [inline cli2/dependencies.ts:7]: get-uri-wrapper works a facade so HTTP types can be mocked.

File: package.json [hunk 11]
```diff
@@ -18,6 +18,7 @@
     "ncc-morphir": "ncc build cli/morphir.js -o dist/morphir",
     "ncc-morphir-server": "ncc build cli/morphir-elm-develop.js -o dist/morphir-server",
     "build": "gulp && npm run ncc-morphir && npm run ncc-morphir-server && npx jest && gulp test",
+    "build-cli2": "gulp buildCLI2",
```
Comment (reebayroo) [inline package.json:21]: To speed up process of running integration tests

**Review by DamianReeves — APPROVED** [review]
LGTM

**Pull request merged**